tigroo31/its-client
# Add cpm

**How to test:**

0- Run the the mosquitto server 

```
$ sudo systemctl start mosquitto
...
```

1- Subscribe to the its fake server:

```
$ mosquitto_sub -v -h 127.0.0.1 -p 1883 -t "#"
...
```

2- Run the copycat client:

```
cargo run --color=always --bin its-client --manifest-path /home/kvmk8371/workspace/its-client/rust/its-client/Cargo.toml -- -H 127.0.0.1 -P 1883 --mqtt-client-id copycat -t 5GCroCo/outQueue
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/its-client -H 127.0.0.1 -P 1883 --mqtt-client-id copycat -t 5GCroCo/outQueue`
Log is written to log/its-client_rCURRENT.log
...
```

3- Publish 3 cpm's, a basic, a standard and a fully used:

```
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/cpm/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type":"cpm","origin":"self","version":"1.0.0","source_uuid":"uuid1","timestamp":1574778515425,"message":{"protocol_version":1,"station_id":12345,"message_id":14,"generation_delta_time":65535,"management_container":{"station_type":5,"reference_position":{"latitude":426263556,"longitude":-82492123,"altitude":800001},"confidence":{"position_confidence_ellipse":{"semi_major_confidence":4095,"semi_minor_confidence":4095,"semi_major_orientation":3601},"altitude":15}},"number_of_perceived_objects":1}}'
kvmk8371@yd-5CD9245WBL:~/workspace/its-client$ mosquitto_pub -t 5GCroCo/outQueue/v2x/cpm/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type":"cpm","origin":"self","version":"1.0.0","source_uuid":"uuid1","timestamp":1574778515425,"message":{"protocol_version":1,"station_id":12345,"message_id":14,"generation_delta_time":65535,"management_container":{"station_type":5,"reference_position":{"latitude":426263556,"longitude":-82492123,"altitude":800001},"confidence":{"position_confidence_ellipse":{"semi_major_confidence":4095,"semi_minor_confidence":4095,"semi_major_orientation":3601},"altitude":15}},"station_data_container":{"originating_vehicle_container":{"heading":180,"speed":1600,"confidence":{"heading":127,"speed":127}}},"sensor_information_container":[{"sensor_id":1,"type":3,"detection_area":{"vehicle_sensor":{"ref_point_id":0,"x_sensor_offset":-20,"y_sensor_offset":20,"vehicle_sensor_property_list":[{"range":5000,"horizontal_opening_angle_start":600,"horizontal_opening_angle_end":600}]}}}],"perceived_object_container":[{"object_id":0,"time_of_measurement":50,"object_confidence":10,"distance":{"x_distance":400,"y_distance":100},"distance_confidence":{"x_distance":102,"y_distance":102},"speed":{"x_speed":1400,"y_speed":500},"speed_confidence":{"x_speed":127,"y_speed":127},"object_ref_point":0}],"number_of_perceived_objects":1}}'
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/cpm/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type":"cpm","origin":"self","version":"1.0.0","source_uuid":"uuid1","timestamp":1574778515425,"message":{"protocol_version":1,"station_id":12345,"message_id":14,"generation_delta_time":65535,"management_container":{"station_type":5,"reference_position":{"latitude":426263556,"longitude":-82492123,"altitude":800001},"confidence":{"position_confidence_ellipse":{"semi_major_confidence":4095,"semi_minor_confidence":4095,"semi_major_orientation":3601},"altitude":15}},"station_data_container":{"originating_vehicle_container":{"heading":180,"speed":1600,"drive_direction":0,"vehicle_length":31,"vehicle_width":18,"longitudinal_acceleration":8,"yaw_rate":4,"lateral_acceleration":2,"vertical_acceleration":1,"confidence":{"heading":127,"speed":127,"vehicle_length":3,"yaw_rate":2,"longitudinal_acceleration":12,"lateral_acceleration":13,"vertical_acceleration":14}},"originating_rsu_container":{"intersection_reference_id":{"road_regulator_id":31,"intersection_id":31470},"road_segment_reference_id":124}},"sensor_information_container":[{"sensor_id":1,"type":3,"detection_area":{"vehicle_sensor":{"ref_point_id":0,"x_sensor_offset":-20,"y_sensor_offset":20,"z_sensor_offset":0,"vehicle_sensor_property_list":[{"range":5000,"horizontal_opening_angle_start":600,"horizontal_opening_angle_end":600,"vertical_opening_angle_start":300,"vertical_opening_angle_end":300}]}}}],"perceived_object_container":[{"object_id":0,"time_of_measurement":50,"object_confidence":10,"distance":{"x_distance":400,"y_distance":100},"distance_confidence":{"x_distance":102,"y_distance":102},"speed":{"x_speed":1400,"y_speed":500},"speed_confidence":{"x_speed":127,"y_speed":127},"object_ref_point":0}],"number_of_perceived_objects":1}}'
$ 
```

4- Check you received 3 cpm's and sent 2:

```
copycat_41470 cpm received_on broker/5GCroCo/outQueue/v2x/cpm/uuid1 12345/65535 at 1630508483891
copycat_41470 cpm received_on broker/5GCroCo/outQueue/v2x/cpm/uuid1 12345/65535 at 1630508574929
copycat_41470 cpm sent_on broker/5GCroCo/inQueue/v2x/cpm/copycat_41470 43815/65535 at 1630508574929
copycat_41470 cpm received_on broker/5GCroCo/outQueue/v2x/cpm/uuid1 12345/65535 at 1630508642521
copycat_41470 cpm sent_on broker/5GCroCo/inQueue/v2x/cpm/copycat_41470 43815/65535 at 1630508642522
```

5- Test that you copy cat the CAM's on your favorite broker.
